devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+# `channel` requirements
+
+## Overview
+
+`channel` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers of data call `channel_pull` to register a `PULL_CALLBACK` function to be called when data becomes available.
+
+Producers of data call `channel_push` to notify the channel that data is available and provide a `PUSH_CALLBACK` to be called when the data has been consumed.
+
+`channel_pull` and `channel_push` can be called from different threads in any order. The callbacks are executed on threadpool threads.
+
+## `channel` internal state
+
+The internal state of the `channel` module can have the following values:
+  - `CHANNEL_EMPTY`: This is the initial state of the `channel`. It means that there is no data available to be pulled. Calling `channel_pull` or `channel_push` will cause the state to transition to `CHANNEL_PUSHING` or `CHANNEL_PULLING` respectively.
+  - `CHANNEL_PUSHED`: This is the state after `channel_push` has been called. A subsequent call to `channel_pull` will schedule the callbacks on the threadpool.
+  - `CHANNEL_PULLED`: This is the state after `channel_pull` has been called. A subsequent call to `channel_push` will the callbacks on the threadpool.
+
+```mermaid
+stateDiagram-v2
+    B: STATE=CHANNEL_EMPTY
+    N: STATE=CHANNEL_PUSHED
+    R: STATE=CHANNEL_PULLED
+    E: STATE=CHANNEL_EMPTY
+
+    [*] --> B: channel_create
+
+    B --> N: channel_push
+    B --> R: channel_pull
+
+    N --> E:  channel_pull
+
+    R --> E: channel_push
+
+    N --> E:  cancel channel_push
+    R --> E: cancel channel_pull
+
+    R --> R: channel_pull
+    N --> N: channel_push
+
+    E --> [*]: channel_destroy
+```
+
+## Reentrancy
+
+`channel_pull` and `channel_push` can be called from callbacks of this module. Since the callbacks are executed on threadpool threads, there is no risk of stack overflow.
```

Comment:
>channel_pull` and `channel_push` can be called from callbacks of this module [](http://example.com/codeflow?start=1&length=76)

Can the callbacks assign NULL to the last reference of the THANDLE(CHANNEL) that is returned from create?

Reply:
That would deadlock